File: midolman/src/main/java/org/midonet/cluster/data/neutron/SecurityGroupApi.java
/*
 * Copyright (c) 2014 Midokura SARL, All Rights Reserved.
 */
package org.midonet.cluster.data.neutron;


import org.midonet.cluster.data.Rule;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.StateAccessException;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.UUID;

public interface SecurityGroupApi {

    /**
     * Create a new security group object
     *
     * @param sg SecurityGroup to create
     * @return SecurityGroup created
     */
    SecurityGroup createSecurityGroup(@Nonnull SecurityGroup sg)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;

    /**
     * Create multiple security group objects.
     *
     * @param sgs List of SecurityGroup objects to create
     * @return List of SecurityGroup objects created
     */
    List<SecurityGroup> createSecurityGroupBulk(
            @Nonnull List<SecurityGroup> sgs)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;

    /**
     * Delete a security group
     *
     * @param id ID of the security group to delete
     */
    void deleteSecurityGroup(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get a security group
     *
     * @param id ID of the security group to fetch
     * @return Security group of the ID provided
     */
    SecurityGroup getSecurityGroup(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get security groups
     *
     * @return All the security groups
     */
    List<SecurityGroup> getSecurityGroups()
            throws StateAccessException, SerializationException;

    /**
     * Update a security group
     *
     * @param id ID of the security group to update
     * @return security group updated
     */
    SecurityGroup updateSecurityGroup(@Nonnull UUID id,
                                             @Nonnull SecurityGroup sg)
            throws StateAccessException, SerializationException;

    /**
     * Create a new security group rule object
     *
     * @param rule SecurityGroupRule to create
     * @return SecurityGroupRule created
     */
    SecurityGroupRule createSecurityGroupRule(
            @Nonnull SecurityGroupRule rule)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;

    /**
     * Create multiple security group rule objects.
     *
     * @param rules List of SecurityGroupRule objects to create
     * @return List of SecurityGroupRule objects created
     */
    List<SecurityGroupRule> createSecurityGroupRuleBulk(
            @Nonnull List<SecurityGroupRule> rules)
            throws StateAccessException, SerializationException,
            Rule.RuleIndexOutOfBoundsException;

    /**
     * Delete a security group rule
     *
     * @param id ID of the security group rule to delete
     */
    void deleteSecurityGroupRule(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get a security group rule
     *
     * @param id ID of the security group rule to fetch
     * @return Security group rule of the ID provided
     */
    SecurityGroupRule getSecurityGroupRule(@Nonnull UUID id)
            throws StateAccessException, SerializationException;

    /**
     * Get security group rules
     *
     * @return All the security group rules
     */
    List<SecurityGroupRule> getSecurityGroupRules()
            throws StateAccessException, SerializationException;

}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/neutron/plugin/SecurityGroupApi.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.neutron.plugin

import java.util
import java.util.UUID
import javax.annotation.Nonnull

import org.midonet.cluster.rest_api.neutron.models.{SecurityGroup, SecurityGroupRule}
import org.midonet.cluster.rest_api.{ConflictHttpException, NotFoundHttpException}

trait SecurityGroupApi {

    /**
     * Create a new security group object
     *
     * @param sg SecurityGroup to create
     * @return SecurityGroup created
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSecurityGroup(@Nonnull sg: SecurityGroup): SecurityGroup

    /**
     * Create multiple security group objects.
     *
     * @param sgs util.List of SecurityGroup objects to create
     * @return util.List of SecurityGroup objects created
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSecurityGroupBulk(@Nonnull sgs: util.List[SecurityGroup])
    : util.List[SecurityGroup]

    /**
     * Delete a security group
     *
     * @param id ID of the security group to delete
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def deleteSecurityGroup(@Nonnull id: UUID)

    /**
     * Get a security group
     *
     * @param id ID of the security group to fetch
     * @return Security group of the ID provided
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSecurityGroup(@Nonnull id: UUID): SecurityGroup

    /**
     * Get security groups
     *
     * @return All the security groups
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSecurityGroups: util.List[SecurityGroup]

    /**
     * Update a security group
     *
     * @param id ID of the security group to update
     * @return security group updated
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def updateSecurityGroup(@Nonnull id: UUID,
                            @Nonnull sg: SecurityGroup): SecurityGroup

    /**
     * Create a new security group rule object
     *
     * @param rule SecurityGroupRule to create
     * @return SecurityGroupRule created
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSecurityGroupRule(@Nonnull rule: SecurityGroupRule): SecurityGroupRule

    /**
     * Create multiple security group rule objects.
     *
     * @param rules util.List of SecurityGroupRule objects to create
     * @return util.List of SecurityGroupRule objects created
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def createSecurityGroupRuleBulk(@Nonnull
                                    rules: util.List[SecurityGroupRule])
    : util.List[SecurityGroupRule]

    /**
     * Delete a security group rule
     *
     * @param id ID of the security group rule to delete
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def deleteSecurityGroupRule(@Nonnull id: UUID)

    /**
     * Get a security group rule
     *
     * @param id ID of the security group rule to fetch
     * @return Security group rule of the ID provided
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSecurityGroupRule(@Nonnull id: UUID): SecurityGroupRule

    /**
     * Get security group rules
     *
     * @return All the security group rules
     */
    @throws(classOf[ConflictHttpException])
    @throws(classOf[NotFoundHttpException])
    def getSecurityGroupRules: util.List[SecurityGroupRule]
}